### DISCUSSION TITLE
What means SUBSCHEMA_TYPE_CHANGED?

### DISCUSSION THREAD

**Discussion by MrDiba** [Apicurio/apicurio-registry#5338] [Q&A]
I'm experimenting with Apicurio registry and I have created an JSON artifact in which I have inserted multiple versions. Looking like this:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "User",
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "favorite_number": {
      "type": [
        "integer",
        "null"
      ]
    },
    "favorite_color": {
      "type": [
        "string",
        "null"
      ]
    }
  },
  "required": [
    "name"
  ]
}
```
During one of the versions I didn´t copy the first `{` resulting in an incorrect JSON, but it accepted it.

Now if I want to create an new version to create an valid JSON again I receive an `SUBSCHEMA_TYPE_CHANGED` and it doesn´t allow me to create an new version.

I tried searching the documentation and the code on what ` SUBSCHEMA_TYPE_CHANGED` means but I could not figure it out.

Maybe someone can enlighten me on what the error means and how I can fix it?

**Comment by jsenko** [comment]
> During one of the versions I didn´t copy the first { resulting in an incorrect JSON, but it accepted it.

Can you give me the content so I can reproduce it? What were your rule settings?

Also your latest version content and the schema that is causing the error would help, as I need two schemas to see what has caused the compatibility check failure. I guess that it means that somewhere value of `"type"` in a property schema has changed. It is also possible that you've added a `"type"` where there wasn't one before.

**Reply by EricWittmann** [reply]
Also the full response from the REST API call that is failing would be useful.  That should have more information in it than just `SUBSCHEMA_TYPE_CHANGED` hopefully...

**Reply by jsenko** [reply]
There is also an open issue to improve the error messages https://github.com/Apicurio/apicurio-registry/issues/5066 , so we're tracking this problem.

**Comment by MrDiba** [comment]
Steps to reproduce:

1. Open the Apicurio Registry UI
2. Global rules in the UI are as follow: 
3. Validity rule: Full, Compatibility rule: Backward
4.  Under explore click Create artifact
5. For Type select "JSON Schema" and click next
6. Leave the Artifact Metadata empty and click next
7. Version Number "v0.1" and Content:
```
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "User",
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "favorite_number": {
      "type": [
        "integer",
        "null"
      ]
    },
    "favorite_color": {
      "type": [
        "string",
        "null"
      ]
    }
  },
  "required": [
    "name"
  ]
}
```
8. The first '{' is missing for the purpose of this demonstration, click next
9. Leave the Version Metadata empty and click Create
10. The artifact specific rules, Validity rule, Compatibility rule and Integrity rule are not enabled (some are globaly)
11. Click on "Versions" select v0.1 and check the content; It contains an invalid JSON missing the first '{'
12. Go back to "Versions" and select Create version
13. Create version "v0.2" with the following content:
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "User",
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "favorite_number": {
      "type": [
        "integer",
        "null"
      ]
    },
    "favorite_color": {
      "type": [
        "string",
        "null"
      ]
    }
  },
  "required": [
    "name"
  ]
}
```
14. Click create and observe the `SUBSCHEMA_TYPE_CHANGED` error

Taking an look at the Network tab in the developer tools:

I see an http `409 Conflict` response on the post made when creating an new version.

The request is:

```
content	{…}
    content	'{\n "$schema": "http://json-schema.org/draft-07/schema#",\n "title": "User",\n "type": "object",\n "properties": {\n "name": {\n "type": "string"\n },\n "favorite_number": {\n "type": [\n "integer",\n "null"\n ]\n },\n "favorite_color": {\n "type": [\n "string",\n "null"\n ]\n }\n },\n "required": [\n "name"\n ]\n}'
     contentType	"application/json"
version	"v0.2"
```

And the answer is:

```
causes	[ {…} ]
    0	    Object { description: "SUBSCHEMA_TYPE_CHANGED", context: "" }
            description	"SUBSCHEMA_TYPE_CHANGED"
            context	""
     detail	"RuleViolationException: Incompatible artifact: e88d27fb-8c89-4d6d-bd8d-984effb965d3 [JSON], num of incompatible diffs: {1}, list of diff types: [SUBSCHEMA_TYPE_CHANGED at ] Causes: SUBSCHEMA_TYPE_CHANGED"
     title	"Incompatible artifact: e88d27fb-8c89-4d6d-bd8d-984effb965d3 [JSON], num of incompatible diffs: {1}, list of diff types: [SUBSCHEMA_TYPE_CHANGED at ] Causes: SUBSCHEMA_TYPE_CHANGED"
     status	409
     name	"RuleViolationException"
```
